distributed/comm/ucx.py

```diff
@@ -133,41 +133,42 @@ def peer_address(self) -> str:
             msg, serializers=serializers, on_error=on_error
         )  # TODO: context=
         nframes = struct.pack("Q", len(frames))
-        await self.ep.send_obj(nframes, sys.getsizeof(nframes))  # send number of frames
+        await self.ep.send_obj(nframes)
 
         for frame in frames:
-            if isinstance(frame, memoryview):
-                # TODO: UCX-PY #27
-                frame = frame.tobytes()
-            size = sys.getsizeof(frame)
-            await self.ep.send_obj(frame, size)
+            await self.ep.send_obj(frame)
         return sum(map(nbytes, frames))
 
     async def read(self, deserializers=None):
         resp = await self.ep.recv_future()
-        # XXX: this breaks things, e.g. test_ucx_specific
-        # dummy = b'0' * 8
-        # resp = await self.ep.recv_obj(dummy, 41)
         obj = ucp.get_obj_from_msg(resp)
         n_frames, = struct.unpack("Q", obj)
 
-        # Notes:
-        # 1. Eventually, ucp_msg will be our abstraction over GPU vs. CPU
-        #    memory. We won't need to worry about checking for the destination
-        #    here. The message object will have a reference to the region of
-        #    memory. Downstream of us (say from _frames) will deserialize
-        #    appropriately, based on whether it's a GPU or CPU object.
-        # 2. We will still deserialize the header early, to check the *length*
-        #    of the next message. This lets us use the faster `recv_obj` to
-        #    read the next nbytes.
-        _header_start = b'\x83\xa7headers'
-
         frames = []
-        msg = {}
+        # gpu_inbound and sizes are for seeing if we can
+        # 1. Take a fastpath to recv a known-length object
+        # 2. Take a fast-fastpath to recv into GPU memory.
+        # see peek for more.
+        gpu_inbound = 0
+        sizes = []
 
         for i in range(n_frames):
-            resp = await self.ep.recv_future()
+            if sizes:
+                this_size = sizes.pop()
+                # XXX: when do we get multiple keys here? Non-contiguous?
+                resp = await self.ep.recv_obj(this_size, cuda=bool(gpu_inbound))
+                # prepare for the next (header) recv
+                if gpu_inbound:
+                    gpu_inbound -= 1
+            else:
+                resp = await self.ep.recv_future()
             frame = ucp.get_obj_from_msg(resp)
+            if should_peek(frame):
```

Comment:
I wonder if we can maybe bake the gpu/cpu target information into an initial frame at the beginning of the message.

```diff
          frames = await to_frames(
              msg, serializers=serializers, on_error=on_error
          )
+         gpu_frame = b''.join([struct.pack(..., is_gpu_frame(frame) for frame in frames])
+         frames = [gpu_frame] + frames
          nframes = struct.pack("Q", len(frames))
```

Reply:
Hmm, that does sound much nicer. I'll take a look now.